Erlang implementation

potatosalad opened the pull request:
I was curious how Erlang would compare using binary pattern matching.  It's patterned roughly after the Elixir implementation.

**Update 2015-05-20:** Added the option to run the the Erlang implementation using unsafe, binary, or regex mode.  Unsafe reads the entire dataset into memory, which is the fastest, but may not work on larger datasets.  The other modes read each file line-by-line instead.

**Update 2015-05-15:** Using [binary:match/2,3](http://www.erlang.org/doc/man/binary.html#match-2) and [binary:part/2,3](http://www.erlang.org/doc/man/binary.html#part-2) dropped the timing on my machine from 8.9 seconds to 2.5 seconds.

*Hardware:* MacBook Pro 2.6GHz i7 (quad core) with 16GB RAM and PCIe SSD (very similar to the hardware in the [part 2 blog post](http://blog.dimroc.com/2015/05/07/etl-language-showdown-pt2/))

<table>
  <tr>
    <td>Erlang w/ binary pattern matching (unsafe)</td>
    <td><del>8.9s</del> <ins>2.5s</ins></td>
  </tr>
  <tr>
    <td>Erlang w/ binary pattern matching (binary)</td>
    <td>4.4s</td>
  </tr>
  <tr>
    <td>Erlang w/ regular expression (regex)</td>
    <td>6.6s</td>
  </tr>
  <tr>
    <td>Ruby w/ Celluloid</td>
    <td>31.6s</td>
  </tr>
  <tr>
    <td>Ruby w/ grosser/parallel</td>
    <td>8.6s</td>
  </tr>
  <tr>
    <td>Elixir</td>
    <td>13.3s</td>
  </tr>
  <tr>
    <td>Golang w/ Substring</td>
    <td>8.6s</td>
  </tr>
</table>

While playing around with the implementation, I wrote three other binary pattern matching implementations which used different strategies (these aren't included in this pull request):

<table>
  <tr>
    <td>Erlang w/ ets (<a href="https://github.com/potatosalad/etl-language-comparison/tree/erlang_ets" target="_blank">erlang_ets branch</a>)</td>
    <td><del>8.9s</del> <ins>2.5s</ins></td>
  </tr>
  <tr>
    <td>Erlang w/ <a href="https://github.com/basho/riak_pipe" target="_blank">basho/riak_pipe</a> (<a href="https://github.com/potatosalad/etl-language-comparison/tree/erlang_riak_pipe" target="_blank">erlang_riak_pipe branch</a>)</td>
    <td><del>30.4s</del> <ins>15.2s</ins></td>
  </tr>
  <tr>
    <td>Erlang w/ single process dict (<a href="https://github.com/potatosalad/etl-language-comparison/tree/erlang_single_process" target="_blank">erlang_single_process branch</a>)</td>
    <td><del>33.4s</del> <ins>12.6s</ins></td>
  </tr>
</table>

The riak_pipe implementation is interesting as it could be run across multiple nodes and could potentially scale horizontally to handle larger datasets better than some of the other implementations.

josevalim commented:
Really nice! It should not be hard to adapt Elixir's implementation after this pattern too if it is "allowed" (I thought the regex was a hard requirement).

josevalim commented:
Also worth pointing out the current implementation loads all file into memory. Which may be ok depending on the size of the dataset.

potatosalad commented:
@josevalim Yeah, I probably drifted away from the original intent of this project.  I saw that the go implementation used substring pattern matching, so I thought binary pattern matching might also be allowed.  Plus, it was too much fun to find out how fast I could get it to go.

I made a few changes so the implementation could be run in 1 of 3 modes:

1. `unsafe` uses my initial "cheating" implementation :grin: (~2.5s)
2. `binary` uses binary pattern matching with `file:read_line/1` (~4.4s)
3. `regex` uses regular expressions with `file:read_line/1` (~6.6s)

josevalim commented:
@potatosalad definitely. I have learned quite a bit with your experiment too. :D I have even open up a thread on Erlang mailing list asking about binary:split/3 which seems to be the bottleneck in the Elixir example...

dimroc commented:
I'm late to the party but great job optimizing the erlang implementation. We've had a lot of language submissions that are optimized far better than my original submissions. But comparisons are becoming apples to oranges at this point because of inconsistencies across the languages (regex, substring, ascii, unicode, map reduction, etc, loading everything into memory as opposed to streaming).

Great job with the `binary:split/3` discovery, I noticed the elixir implementation improved dramatically. I will merge all of these with master but write about the differences.

josevalim commented:
@dimroc let me know if you would like those changes to be ported to Elixir. I would really like to follow Erlang footsteps in here otherwise it will generate a whole "Elixir is 5x slower than Erlang" which is certainly not true. :)

Thanks for merging this.

dimroc commented:
Sure @josevalim that would be great. Can you do me two favors though?
1. Keep the original implementation, similar to how @potatosalad has three modes.
2. Help me explain the difference between the implementations :smile:. I feel like I only have a superficial grasp on the the powers of `binary/match` and `binary/split` and would love to have a better write up in the next blog post.

You can see the early (early) draft of that post here (minus bar charts): https://github.com/dimroc/blog/blob/source/_drafts/2015-08-31-etl-language-showdown-pt3.md

Hope I'm not asking too much.

josevalim commented:
Definitely. I will work on that and hopefully I have something on monday. :)

josevalim commented:
Btw, I can explain 2 right now. :)

1. The move from regex to `binary:split/2` makes things faster because, similar to other languages, there is a cost when running regular expressions which can be removed by using operations on binaries

2. After moving to `binary:split/2`, you could further speed up things by precompiling the pattern you want to split on. So if you are going to split or match "knicks", you can precompile the pattern before hand and it will improve performance on tight loops (as you skip precompiling it on every match/split). @potatosalad was particularly smart here because he calculated all uppercase and lowercase permutations of knicks into a pattern

3. Then @potatosalad found out that he could get even better performance by not using split and using `binary:match/2`. `binary:split/2` in Erlang is implemented in two steps, first it uses `binary:match/2` to get all positions and then it uses `binary:part/2` to extract a binary from those positions. However, imagine that you have a CSV row "foo,bar,baz,bat" and you are interested only on the columns bar and bat. Even if you don't use foo and baz, `binary:split/2` will convert those to binaries. @potatosalad found that he could avoid the extra cost by not converting foo and baz to binaries.

4. Bonus: @potatosalad also realized that `binary:split/2` could be faster if implemented in C. Remember when I said that `binary:split/2` is `binary:match/2` + `binary:part/2`? Because `binary:part/2` needs to check the boundaries and `binary:split/2` is implemented in Erlang, it means that, even though `binary:split/2` returns valid boundaries, `binary:part/2` called by split needs to check those boundaries all the time too which adds an extra cost. @potatosalad was extra awesome by sending a PR to Erlang OTP.

@potatosalad, feel free to correct me if I got anything wrong. :)

potatosalad commented:
@josevalim After reviewing the code again, that all sounds correct.  Thank you for the explanation, it's much more informative than I wrote in the description above.  Also, sorry to have caused you extra work so things wouldn't appear to be "Elixir is 5x slower than Erlang" :grimacing:

Interesting side notes that will hopefully be part of OTP 19:

I ran the [original elixir implementation which used String.split/2](https://github.com/dimroc/etl-language-comparison/blob/127c76f25211fb4ca7e7faa8c5baf6b045a4943a/elixir/lib/map_actor.ex#L8) against the `binary:split/2,3` pull request at erlang/otp#771 and the time dropped from ~12 seconds to ~8 seconds on my machine.  So the original, and arguably more maintanable, version of the code should see a 1-2x speedup without any modifications if I'm able to get my OTP PR accepted for OTP 19.

I was also curious how much a speedup a split based solution running with the OTP PR would provide for Elixir.  If you're interested it's on this branch: [potatosalad/etl-language-comparison@match_and_split](https://github.com/potatosalad/etl-language-comparison/tree/match_and_split).

<table>
  <tr>
    <td>Elixir (binary) OTP 18</td>
    <td>2.9s</td>
  </tr>
  <tr>
    <td>Elixir (regex) OTP 18</td>
    <td>7.9s</td>
  </tr>
  <tr>
    <td>Elixir <a href="https://github.com/potatosalad/etl-language-comparison/blob/match_and_split/elixir/lib/split_map_actor.ex">(split)</a> OTP 18</td>
    <td>3.7s</td>
  </tr>
  <tr>
    <td>Elixir (binary) OTP 18 patched</td>
    <td>2.9s</td>
  </tr>
  <tr>
    <td>Elixir (regex) OTP 18 patched</td>
    <td>6.8s</td>
  </tr>
  <tr>
    <td>Elixir <a href="https://github.com/potatosalad/etl-language-comparison/blob/match_and_split/elixir/lib/split_map_actor.ex">(split)</a> OTP 18 patched</td>
    <td>2.5s</td>
  </tr>
</table>

The [regex implementation](https://github.com/potatosalad/etl-language-comparison/blob/match_and_split/elixir/lib/regex_map_actor.ex) looks like it benefits from the `binary:split/2,3` patch as well as the [split implementation](https://github.com/potatosalad/etl-language-comparison/blob/match_and_split/elixir/lib/split_map_actor.ex).  We're now the same speed as the Erlang implementation I posted in this PR.

However, while fooling around with the implementations, I stumbled upon an even faster method:

<table>
  <tr>
    <td>Elixir <a href="https://github.com/potatosalad/etl-language-comparison/blob/match_and_split/elixir/lib/match_map_actor.ex">(match)</a></td>
    <td>1.1s</td>
  </tr>
</table>

Therefore, while there is currently no similar implementation in Erlang, let the record show that Elixir is roughly 2-3x faster than Erlang :wink:

Also, I think 1.1 seconds is currently the fastest runtime for any of the languages currently posted on this repository (it beats the [Rust implementation](https://github.com/dimroc/etl-language-comparison/pull/17) by roughly 1-2x).

dimroc commented:
This is all great information, so I'm thinking it would be great to capture details of the implementations in a README.md in the erlang/ folder. @potatosalad, would you be up to the task? You seem the most qualified :smile:
